### DISCUSSION TITLE
Print Help, add additional help, and the commander instance

### DISCUSSION THREAD

**Discussion by jmcdo29** [jmcdo29/nest-commander#16] [Ideas]
I'm thinking about exporting an `@InjectCommander()` decorator for injecting the `commander` instance to allow devs to call `this.commander.help()` if necessary. And while working with Help stuff, I'm thinking about creating a few new decorators for adding help to the command [as described here](https://www.npmjs.com/package/commander#automated-help)

Probably some decorators like 

```ts
@HelperText('beforeAll' | 'before' | 'after' | 'afterAll')
printHelp() {
  return 'String that should be used for help';
}
```

All of these options could be used, and would be added to commander appropriately.

Might even add in ways to change the help flag, but that'll be a stretch from the current implementation. Anyone got any thoughts?

**Comment by Zyles** [comment]
Any progress on this? I don't see a way to display the help of a command when ran without options. Only the top level help can be shown. When running `node dist/console`

**Comment by smolinari** [comment]
@Zyles - It looks like this was added. 

https://jmcdo29.github.io/nest-commander/docs/features/commander/#adding-custom-help

Scott

**Reply by Zyles** [reply]
Tried this today but not sure how to invoke this.

When user passes no arguments on my command I want to print the commander default --help text.

`$ node dist/console mycommand`

Should output:
```
Usage: console mycommand [options] [command]

Mycommand description text.

Options:
  -h, --help                 display help for command

Commands:
  import:products [options]  Run products importer.
```

But right now it outputs:
`No command provided. See --help for commands.`

With the code:
```
@Command({
  name: 'mycommand',
  description: 'Mycommand description text.',
  subCommands: [],
})
export class MyCommand implements CommandRunner {
  async run(passedParam: string[], options: any): Promise<void> {
    console.log(`No command provided. See --help for commands.`)
    exit()
  }
}
```

How do I implement `@Help()` for this?

**Reply by mikehaertl** [reply]
@Zyles You can inject the command instance in the constructor. The following kind of works for me. My only problem is that `this.command.help()` always creates the top level command help. 

@jmcdo29 Do you maybe have an idea why that is or how to fix this?

```
$ pnpm cli user
Usage: cli [options] [command]

Options:
  -h, --help      display help for command

Commands:
  user            manage app users
  help [command]  display help for command
```

I would like it to create the help for the `user` command including the subcommands, just like it does when calling `help user`: 

```
$ pnpm cli help user
Usage: cli user [options] [command]

manage app users

Options:
  -h, --help  display help for command

Commands:
  list|l      list all app users
```

```typescript
import { Command, CommandRunner, InjectCommander } from 'nest-commander';
import { Command as Commander } from 'commander';
import { ListUserCommand } from './list-user.command';

@Command({
  name: 'user',
  description: 'manage app users',
  subCommands: [ListUserCommand],
})
export class UserCommand extends CommandRunner {
  constructor(@InjectCommander() private commander: Commander) {
    super();
  }

  async run(inputs: string[], options: Record<string, any>) {
    this.commander.outputHelp();
  }
}
```

```typescript
import { CommandRunner, SubCommand } from 'nest-commander';
import { UserService } from '../users/user.service';

@SubCommand({
  name: 'list',
  aliases: ['l', 'li'],
  description: 'list all app users',
})
export class ListUserCommand extends CommandRunner {
  constructor(private userService: UserService) {
    super();
  }

  async run(inputs: string[], options: Record<string, any>) {
    process.stdout.write('Listing users!\n');
    const users = await this.userService.findAll();
    console.log(users);
  }
}
```

**Reply by mikehaertl** [reply]
I found a way to show the correct command help. It's not super nice but it works:

```typescript
  async run(inputs: string[], options: Record<string, any>) {
    console.log(this.commander.commands);
    const cmd = this.commander.commands.find(c => c.name() === 'user');
    cmd?.outputHelp();
  }
```

**Reply by jmcdo29** [reply]
In nest-commander@3.0.0 there's also now the `CommandRunner#command` member, so you should be able to access the specific command rather than having to inject the full commander instance and then finding off of a name

**Reply by mikehaertl** [reply]
Oh, that's cool. Works like a charm:

```typescript
import { Command, CommandRunner } from 'nest-commander';
import { ListUserCommand } from './list-user.command';

@Command({
  name: 'user',
  aliases: ['u', 'us'],
  description: 'manage app users',
  subCommands: [ListUserCommand],
})
export class UserCommand extends CommandRunner {
  async run(inputs: string[], options: Record<string, any>) {
    this.command.outputHelp();
  }
}
```

Is this already documented somewhere?

**Reply by jmcdo29** [reply]
[Documentation was updated when 3.0.0 was released](https://nest-commander.jaymcdoniel.dev/docs/features/commander/#getting-the-current-command)
